theano/tensor/nnet/nnet.py

```diff
@@ -54,35 +54,30 @@ class SoftmaxWithBias(gof.Op):
 
     """
 
-    nin = 2
+    nin = 3
     nout = 1
     __props__ = ()
 
-    def make_node(self, x, b):
+    def make_node(self, x, b, axis=-1):
         x = tensor.as_tensor_variable(x)
         b = tensor.as_tensor_variable(b)
-        if x.type.ndim != 2 \
-                or x.type.dtype not in tensor.float_dtypes:
-            raise ValueError('x must be 2-d tensor of floats')
+        axis = tensor.as_tensor_variable(axis)
```

Comment:
> It could make some things much easier, for instance shape inference.

... and optimization (we will want to use cuDNN if `axis==1` and `x.ndim in (4,5)`), and replacing it with an expression graph that uses `T.sum()` and `T.max()` (if ever the compiled expression graph becomes as fast as the manual C code).